src/DynamicExpresso.Core/Parsing/Parser.cs

```diff
@@ -1386,7 +1386,8 @@ private static bool IsDynamicType(Type type)
 
 		private static bool IsDynamicExpression(Expression instance)
 		{
-			return instance != null && instance.NodeType == ExpressionType.Dynamic;
+			return instance != null &&
+				(instance.NodeType == ExpressionType.Dynamic || instance.Type == typeof(object));
```

Comment:
There are no other ways to understand if the expression is declared as `dynamic`?
From outside just checking if it is an object seems to be wrong. But I admit that I don't see any real problem.

Maybe what I can just think of is what happen if you add some tests where we use an object? For example writing something like:
```
		public class NotDynamicProperties
        {
			public object Foo => 100;
			public object Bar() => 100;
        }

		[Test]
		public void Math_with_not_dynamic_properties_of_class()
		{
			var notDynamicProperties = new NotDynamicProperties();
			var interpreter = new Interpreter();
			var del = interpreter.ParseAsDelegate<Func<NotDynamicProperties, int>>("d.Foo+d.Bar()", new[] { "d" });
			var r = del.Invoke(dynamicProperties);
			// TODO add an assert to check exception
		}
```

I would expect a ParseException. Correct?

Reply:
This this is effective the same as the test I've included, since `public object Foo` and `public dynamic Foo` results in the same IL afaik. With the change proposed in this pull request, your suggested test won't throw.

Reply:
To be clear, in "real" C#, the dynamic keyword instructs the compiler to wire up the late biding code instead of the normal path which would simply not compile. Example : 
![image](https://user-images.githubusercontent.com/840042/122826656-c7ee5480-d2b9-11eb-8f2b-0b4d993e4ade.png)
What this pull requests enable is late biding on objects without the dynamic keyword being required, which is a tiny deviation from C#. 
The real problem is that when doing reflection on a type like `DynamicProperties` , the properties and methods that are declared as `dynamic` are seen as having an `object` type. If we have an instance of that class, then we can get the actual type for those.

Reply:
Yes, my concern is that this kind of "late binding" can introduce some bad side effect.

Maybe we can enable it using a feature flag? We can add a "ObjectLateBinding" option [here](https://github.com/davideicardi/DynamicExpresso/blob/master/src/DynamicExpresso.Core/InterpreterOptions.cs). Inside `IsDynamicExpression` check `|| instance.Type == typeof(object)` only if `ObjectLateBinding` is set.
And in the tests check the different behaviours depending on the flag.

What do you think?

Reply:
That's probably wise and a very good idea. I'll amend the pull request with those.

Reply:
Done @davideicardi ,
I've modified the test so it asserts an exception is throw when the flag is not enabled